File: 00_Hello_World/Python/src/main.py
import sys


def say_hello(name):
    print(f"Hello, {name}!")


def print_error(msg):
    if msg is not None:
        print(msg)

    print(
        'Use:                                                \n'
        '  hello_world [...args]                             \n'
        '                                                    \n'
        '  without arguments, print "Hello, World!"          \n'
        '                                                    \n'
        '  -n <name> - print "Hello, <name>!"                \n'
        '  -i        - ask <name> and print "Hello, <name>!" \n'
    )


def main():
    argv = sys.argv
    argc = len(argv)
    if argc == 0:
        assert False, "Invalid environment"
    elif argc == 1:
        say_hello("World")
    elif argc == 2:
        if argv[1] != "-i":
            print_error("Unexpected argument on first place, -i expected.")
        else:
            name = input("Input your name: ")
            say_hello(name)
    elif argc == 3:
        if argv[1] != "-n":
            print_error("Unexpected argument on first place, -n expected.")
        else:
            say_hello(argv[2])
    else:
        print_error("Incorrect count of arguments")


if __name__ == '__main__':
    main()

File: 00_Hello_World/Cpp/src/main.cpp
#include <iostream>
#include <string>
#include <cassert>


namespace {
void SayHello(const std::string& name) {
    std::cout << "Hello, " << name << "!" << std::endl;
}


void SayHelloInteractive() {
    std::string name;
    std::cout << "Input your name: " << std::flush;
    std::cin >> name;

    SayHello(name);
}


void PrintError(const std::string& msg = "") {
    if (not msg.empty()) {
        std::cout << msg << std::endl;
    }
    std::cout << "Use:                                                  \n"
                 "  hello_world [...args]                               \n"
                 "                                                      \n"
                 "  without arguments, print \"Hello, World!\"          \n"
                 "                                                      \n"
                 "  -n <name> - print \"Hello, <name>!\"                \n"
                 "  -i        - ask <name> and print \"Hello, <name>!\" \n"
              << std::flush;
}
} // namespace _


/// hello_world [...args]
///
/// without arguments, print "Hello, World!"
///
/// -n <name> - print "Hello, <name>!"
/// -i        - ask <name> and print "Hello, <name>!"
int main(int argc, char* argv[]) {
    switch (argc) {
        case 0:
            assert(false && "");
        case 1:
            SayHello("World");
            break;
        case 2:
            if (argv[1] != std::string{"-i"}) {
                PrintError("Unexpected argument on first place, -i expected.");
                break;
            }
            SayHelloInteractive();
            break;
        case 3:
            if (argv[1] != std::string{"-n"}) {
                PrintError("Unexpected argument on first place, -n expected.");
                break;
            }
            SayHello(argv[2]);
            break;
        default:
            PrintError("Unexpected count of arguments.");
            break;
    }
    return 0;
}
